### PR TITLE
feat(plugin-cc): add agent logout

### PR DESCRIPTION
**Pull request by rarajes2**
# COMPLETES #< [SPARK-558556](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-558556) >

## This pull request addresses

Adds the Agent Station Logout feature

## by making the following changes

Agent is able to Logout from station.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

Tested the station logout on the samples page after logging in with Browser.

### I certified that

- [ ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: docs/samples/contact-center/index.html
Comment (bhabalan) [inline docs/samples/contact-center/index.html]: Prefer using the already present 'hidden' class instead of display: none

File: packages/@webex/plugin-cc/src/cc.ts
Comment (bhabalan) [inline packages/@webex/plugin-cc/src/cc.ts:137]: Add a log for the success case

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (bhabalan) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: The options param can be passed as is to agentService without the need for destructuring

File: packages/@webex/plugin-cc/src/services/AgentService.ts
Comment (bhabalan) [inline packages/@webex/plugin-cc/src/services/AgentService.ts]: These strings can be constants

File: packages/@webex/plugin-cc/src/services/types.ts
Comment (bhabalan) [inline packages/@webex/plugin-cc/src/services/types.ts]: Use a constant for this string

File: packages/@webex/plugin-cc/src/services/types.ts
Comment (bhabalan) [inline packages/@webex/plugin-cc/src/services/types.ts]: Use a constant for this string

File: docs/samples/contact-center/index.html
Comment (rarajes2) [reply docs/samples/contact-center/index.html]: Done

File: packages/@webex/plugin-cc/src/cc.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/cc.ts:137]: `this.agent.stationLogout` is already adding the success log. Adding here will add it twice.

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: Done

File: packages/@webex/plugin-cc/src/services/AgentService.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/AgentService.ts]: Done

File: packages/@webex/plugin-cc/src/services/types.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/types.ts]: Done

File: packages/@webex/plugin-cc/src/services/types.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/types.ts]: Done

File: packages/@webex/plugin-cc/src/cc.ts
Comment (bhabalan) [reply packages/@webex/plugin-cc/src/cc.ts:137]: @rarajes2 , In that case, let's remove the logs from catch

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (bhabalan) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: Add log for failures

**Review by bhabalan — APPROVED** [review]

File: packages/@webex/plugin-cc/src/services/constants.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/constants.ts]: Constant names should be in caps

File: packages/@webex/plugin-cc/src/services/AgentService.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/AgentService.ts]: The return value should be generic right, either the promise resolves or rejects indicating success or failure

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: If 4xx or 5xx failure occurs, don't we need to handle that and reject promise

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: It's better to say Logout Success and Logout Failure. We don't need to add API in the log messages.

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: Let's call it data instead of options

File: packages/@webex/plugin-cc/test/unit/spec/features/Agent.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/test/unit/spec/features/Agent.ts]: We should have tests added for changes in the cc.ts file as well and the same goes for AgentService.ts. And we should cover the failure response test case also

**Review by Kesari3008 — CHANGES_REQUESTED** [review]

**Comment by Kesari3008** [comment]
Build seems to be failing with below error:
[@webex/plugin-cc]: src/WebRTCCalling.ts(8,8): error TS2307: Cannot find module '@webex/calling' or its corresponding type declarations. 

Is this being taken care in any of the PRs?

Please add manual testing details in the PR description

File: packages/@webex/plugin-cc/src/cc.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/cc.ts:137]: Done

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: Done

File: packages/@webex/plugin-cc/src/services/constants.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/constants.ts]: Done

File: packages/@webex/plugin-cc/src/services/AgentService.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/AgentService.ts]: Done

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: Updated the code, now all the logic is in the service, this function simply calls it and returns. It's just one liner now

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: Done

File: packages/@webex/plugin-cc/src/features/Agent.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: Done

File: packages/@webex/plugin-cc/test/unit/spec/features/Agent.ts
Comment (rarajes2) [reply packages/@webex/plugin-cc/test/unit/spec/features/Agent.ts]: Added tests for `cc` and `AgentService`. Removed tests from `Agent` as it's one liner now.

File: packages/@webex/plugin-cc/src/services/constants.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/constants.ts:15]: These could be moved within CC_EVENTS itself but check with Ravi if we would even need CC_EVENTS enum since we are moving the aqm code

File: packages/@webex/plugin-cc/src/services/AgentService.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/AgentService.ts:67]: Would we need to have a try-catch block here? I believe we can just return response as it is that we get from httpRequest.sendRequestWithEvent()

File: packages/@webex/plugin-cc/src/cc.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/cc.ts:129]: Let's create a type for  {logoutReason: string} as Logout similar and use that eveyrwhere

File: packages/@webex/plugin-cc/src/cc.ts
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/cc.ts:138]: Would we need this catch block? I believe return response should suffice

**Review by Kesari3008 — APPROVED** [review]
Approving with few comments. Please ensure to address these comments before merge
Screenshot 2024-11-07 at 8 37 14 PM
Even when logout was successful, timeout error is printing. Please fix this in sample app

**Comment by rarajes2** [comment]
Closing as this PR changes has been taken in Ravi's PR.

**Pull request closed without merging**
